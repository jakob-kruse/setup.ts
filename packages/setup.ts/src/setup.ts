import { z } from "zod";
import { bundleRequire } from "bundle-require";
import { PackageJson } from "./types/package-json";
import type { Plugin, PluginBuilder } from "@/plugin";

const DefineSetupOptionsSchema = z
  .object({
    validateConfig: z.boolean().default(true).optional(),
    validateOptions: z.boolean().default(true).optional(),
  })
  .optional();

export type DefineSetupOptions = z.infer<typeof DefineSetupOptionsSchema>;

export class SetupBuilder {
  private pluginQueue: PluginBuilder[] = [];

  constructor(
    private readonly config: PackageJson,
    private readonly options?: DefineSetupOptions
  ) {}

  public add(plugin: Plugin) {
    this.pluginQueue.push(plugin(this.options));
    return this;
  }

  public async _build() {
    if (this.pluginQueue.length === 0) {
      return this.config;
    }

    const pluginChanges = await Promise.all(this.pluginQueue);

    let newConfig = structuredClone(this.config);

    for (const pluginChange of pluginChanges) {
      newConfig = Object.assign(newConfig, pluginChange);
    }

    return newConfig;
  }
}

export function defineSetup(
  config: PackageJson,
  options?: DefineSetupOptions
): SetupBuilder {
  return new SetupBuilder(config, options);
}

export async function compile(filePath: string) {
  const file = await bundleRequire({
    filepath: filePath,
  });

  if (!file.mod) {
    throw new Error("Could not import package.ts");
  }

  const setupBuilder: SetupBuilder = await file.mod.default;

  return setupBuilder._build();
}
