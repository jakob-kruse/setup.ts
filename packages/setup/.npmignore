src/
.gitignore
tsconfig.json
pnpm-lock.yaml
.turbo
tsup.config.ts
vitest.config.ts