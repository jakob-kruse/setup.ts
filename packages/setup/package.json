{
  "name": "@setup.ts/setup",
  "version": "0.0.1-alpha",
  "description": "Node.js project scaffolding in TypeScript",
  "module": "dist/index.mjs",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "bin": {
    "setup.ts": "dist/cli.js"
  },
  "files": [
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsup",
    "build:watch": "tsup --watch",
    "dev": "vitest dev",
    "test": "vitest run",
    "test:coverage": "vitest --coverage",
    "lint": "eslint {src,test}/**/*.ts",
    "prepublish": "npm run build"
  },
  "keywords": [],
  "author": {
    "name": "Jakob Kruse",
    "email": "contact@jakob-kruse.de"
  },
  "license": "ISC",
  "engines": {
    "node": ">=14"
  },
  "devDependencies": {
    "@types/node": "^17.0.42",
    "c8": "^7.11.3",
    "eslint": "^8.17.0",
    "eslint-config-custom": "workspace:*",
    "tsup": "^6.1.2",
    "typescript": "^4.7.3",
    "vite": "^2.9.12",
    "vite-tsconfig-paths": "^3.5.0",
    "vitest": "^0.14.2"
  },
  "dependencies": {
    "bundle-require": "^3.0.4",
    "cac": "^6.7.12",
    "esbuild": "^0.14.43",
    "zod": "^3.17.3"
  }
}