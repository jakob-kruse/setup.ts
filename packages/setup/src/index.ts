export { definePlugin, defineTemplate } from "@/plugin";
export type {
  SetupPluginBuilder,
  SetupPluginBuilderContext,
  SetupPluginBuilderFn,
  SetupPluginFile,
  SetupPluginResult,
  TemplateFn,
} from "@/plugin";
export { defineSetup } from "@/setup";
export type { SetupBuilder, SetupBuilderOptions } from "@/setup";
