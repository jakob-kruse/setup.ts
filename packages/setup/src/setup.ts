import { SetupPluginBuilder } from "@/plugin";
import {
  PackageDefinition,
  PackageDefinitionSchema,
} from "@/types/package-json";
import { bundleRequire } from "bundle-require";
import { promises as fs } from "fs";
import path from "path";
import { z } from "zod";

const SetupOptionsSchema = z
  .object({
    validatePackageJson: z.boolean().default(true).or(z.undefined()),
  })
  .default({});

export type SetupBuilderOptions = z.infer<typeof SetupOptionsSchema>;

export class SetupBuilder {
  /**
   * Plugins to apply to the package definition
   */
  private plugins: Promise<Partial<PackageDefinition>>[] = [];

  constructor(
    private packageDefinition: PackageDefinition,
    private options: SetupBuilderOptions = {}
  ) {}

  public getOptions(): SetupBuilderOptions {
    return this.options;
  }

  public mergeOptions(options: Partial<SetupBuilderOptions>): this {
    this.options = Object.assign(this.options, options);

    return this;
  }

  public mergePackage(packageJson: Partial<PackageDefinition>): this {
    this.packageDefinition = Object.assign(this.packageDefinition, packageJson);

    return this;
  }

  /**
   * Use a plugin to modify the package definition.
   *
   * @param plugin - The plugin to apply to the package definition
   * @returns
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public use(plugin: any): this {
    this.plugins.push((plugin as SetupPluginBuilder).build());

    return this;
  }

  /**
   * @internal - Do not use this method.
   *
   * Builds the package definition.
   *
   * @returns The built package definition with all plugins applied
   */
  public async _build(): Promise<PackageDefinition> {
    this.mergeOptions(SetupOptionsSchema.parse(this.options));

    if (this.options?.validatePackageJson) {
      await PackageDefinitionSchema.parseAsync(this.packageDefinition);
    }

    if (this.plugins.length === 0) {
      return this.packageDefinition;
    }

    const pluginChanges = await Promise.all(this.plugins);

    for (const pluginChange of pluginChanges) {
      this.mergePackage(pluginChange);
    }

    return this.packageDefinition;
  }

  /**
   * @internal - Do not use this method.
   * @returns A SetupBuilder
   */
  static async fromFile(filePath: string): Promise<SetupBuilder> {
    const resolved = path.resolve(filePath);
    try {
      const fileStat = await fs.stat(resolved);

      if (!fileStat.isFile()) {
        throw new Error(`"${resolved}" is not a file`);
      }
    } catch (error) {
      throw new Error(`"${resolved}" does not exist`);
    }

    const file = await bundleRequire({
      filepath: resolved,
    });

    if (!file.mod) {
      throw new Error(`Could not import "${resolved}"`);
    }

    if (!file.mod.default) {
      throw new Error(`"${resolved}" does not export a default`);
    }

    return await file.mod.default;
  }
}

/**
 * Define a project setup.
 *
 * @param packageDefinition - The package.json data.
 * @param options - The options to define the setup.
 * @returns The setup builder.
 */
export function defineSetup(
  packageDefinition: PackageDefinition,
  options?: SetupBuilderOptions
): SetupBuilder {
  return new SetupBuilder(packageDefinition, options);
}
