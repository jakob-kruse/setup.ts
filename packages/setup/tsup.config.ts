import { defineConfig } from "tsup";

export default defineConfig((options) => ({
  entry: ["src/index.ts", "src/cli.ts"],
  minify: !options.watch,
  splitting: false,
  format: ["cjs", "esm"],
  dts: true,
  sourcemap: true,
  clean: true,
}));
