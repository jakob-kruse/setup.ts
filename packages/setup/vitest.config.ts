import { defineConfig } from "vitest/config";
import tsconfigPaths from "vite-tsconfig-paths";

export default defineConfig({
  test: {
    mockReset: true,
    clearMocks: true,
  },
  plugins: [tsconfigPaths()],
});
