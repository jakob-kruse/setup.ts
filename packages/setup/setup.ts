import { defineSetup } from '@setup.ts/setup';

export default defineSetup({
  "name": "setup",
  "version": "0.0.1",
  "devDependencies": {
    "@setup.ts/setup": "latest"
  }
});