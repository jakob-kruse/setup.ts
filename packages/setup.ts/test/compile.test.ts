import { describe, it, expect } from 'vitest';
import { compile } from '@/setup';

describe('Compilation', () => {
  describe('basic', () => {
    it('should compile sync', async () => {
      const fixturePath = 'test/fixtures/compilation/basic.ts';
      const config = await compile(fixturePath);

      expect(config).toMatchSnapshot();
    });
  });

  describe('plugins', () => {
    it('should deep merge with the plugin result', async () => {
      const fixturePath = 'test/fixtures/compilation/plugin.ts';
      const config = await compile(fixturePath);

      expect(config).toMatchSnapshot();
    });

    it('should be chaining plugins', async () => {
      const fixturePath = 'test/fixtures/compilation/plugin-chaining.ts';
      const config = await compile(fixturePath);

      expect(config).toMatchSnapshot();
    });
  });
});
