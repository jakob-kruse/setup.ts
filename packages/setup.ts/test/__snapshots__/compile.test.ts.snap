// Vitest Snapshot v1

exports[`Compilation > basic > should compile sync 1`] = `
{
  "description": "test-desc",
  "name": "test-name",
  "version": "1.0.0",
}
`;

exports[`Compilation > plugins > should be chaining plugins 1`] = `
{
  "bugs": {
    "email": "test@test.de",
  },
  "description": "changed by test plugin 2",
  "name": "test-name",
  "version": "1.0.0",
}
`;

exports[`Compilation > plugins > should deep merge with the plugin result 1`] = `
{
  "description": "test-desc",
  "name": "test-name",
  "version": "1.1.2",
}
`;
